### PR TITLE
Flask Restx API

### LINKED ISSUE #564: REST firmware data field bug
The data fields of `/rest/firmware` (PUT) are not checked enough. In case the "requested_analysis_systems" are provided as string instead of list, the scheduler crashes with this error:

```
   File "/home/jstucke/git/github/FACT_core/src/scheduler/Analysis.py", line 417, in _get_cumulative_remaining_dependencies
    return {
  File "/home/jstucke/git/github/FACT_core/src/scheduler/Analysis.py", line 420, in <setcomp>
    for dependency in self.analysis_plugins[plugin].DEPENDENCIES
KeyError: 'l'
```

### PR DESCRIPTION
**Pull request by any333**
resolves #564

### CODE REVIEW

File: src/web_interface/rest/rest_firmware.py [hunk 1]
```diff
@@ -18,68 +18,40 @@
...
-
-    def _get_without_uid(self):
+@api.route('', doc={'description': ''})
+class RestFirmwareGetWithoutUid(RestFirmwareBase):
+    @roles_accepted(*PRIVILEGES['view_analysis'])
+    @api.doc(responses={200: 'Success', 400: 'Unknown file object'})
+    def get(self):
+        '''
+        Browse the database
+        List all available firmwares in the database
```
Comment (jstucke) [inline src/web_interface/rest/rest_firmware.py:53]: firmware is not countable (like software): firmwares -> firmware

Comment (jstucke) [reply src/web_interface/rest/rest_firmware.py:53]: Documentation of optional URL arguments (offset, limit, query, recursive and inverted) is missing (see https://github.com/fkie-cad/FACT_core/wiki/Rest-API#get)

File: src/web_interface/rest/rest_firmware.py [hunk 2]
```diff
@@ -106,7 +78,54 @@ def _get_parameters_from_request(request_parameters):
...
+        with ConnectTo(InterComFrontEndBinding, self.config) as intercom:
+            intercom.add_analysis_task(firmware_object)
+        data.pop('binary')
+
+        return dict(uid=firmware_object.uid)
+
+
+@api.route('/<string:uid>',
+           doc={'description': '',
+                'params': {'uid': 'File UID'}
```
Comment (jstucke) [inline src/web_interface/rest/rest_firmware.py:118]: 'File UID' -> 'Firmware UID'

File: src/web_interface/rest/rest_firmware.py [hunk 3]
```diff
@@ -106,7 +78,54 @@ def _get_parameters_from_request(request_parameters):
...
+                }
+           )
+class RestFirmwareGetWithUid(RestFirmwareBase):
+    @roles_accepted(*PRIVILEGES['view_analysis'])
+    @api.doc(responses={200: 'Success', 400: 'Unknown file object'})
```
Comment (jstucke) [inline src/web_interface/rest/rest_firmware.py:123]: 'Unknown file object' -> 'Unknown firmware' or 'Unknown UID'

File: src/web_interface/rest/rest_firmware.py [hunk 4]
```diff
@@ -106,7 +78,54 @@ def _get_parameters_from_request(request_parameters):
...
+    def get(self, uid):
+        '''
+        Request a specific file
+        Get a specific file by providing the uid of the corresponding firmware
```
Comment (jstucke) [inline src/web_interface/rest/rest_firmware.py:127]: This is kind of misleading. You don't really get a "file" in the result. It is a firmware entry from the database (converted to JSON).

Comment (jstucke) [reply src/web_interface/rest/rest_firmware.py:127]: This is also missing the documentation of an optional URL parameter: "summary"

File: src/web_interface/rest/rest_firmware.py [hunk 5]
```diff
@@ -120,23 +139,45 @@ def _get_with_uid(self, uid):
...
+        analysis = firmware.processed_analysis
+        return dict(meta_data=meta, analysis=analysis)
 
-    def _put_with_uid(self, uid):
+    @roles_accepted(*PRIVILEGES['submit_analysis'])
+    @api.expect(firmware_model)
+    def put(self, uid):
+        '''
+        Upload a firmware
+        The HTTP body shall contain a json document of the structure shown below
```
Comment (jstucke) [inline src/web_interface/rest/rest_firmware.py:153]: shall -> must
shall sounds very ... archaic (I know it is like this in the REST docs, too. We really need to revise them. Maybe you can update it there, too)

File: src/web_interface/rest/rest_firmware.py [hunk 6]
```diff
@@ -106,7 +78,54 @@ def _get_parameters_from_request(request_parameters):
             raise ValueError('Inverted flag can only be used with recursive')
         return query, recursive, inverted, offset, limit
 
-    def _get_with_uid(self, uid):
+    @roles_accepted(*PRIVILEGES['submit_analysis'])
+    @api.expect(firmware_model)
+    def put(self):
+        '''
+        Update existing firmware analysis
+        You can use this endpoint to update a firmware analysis which is already existing
```
Comment (jstucke) [inline src/web_interface/rest/rest_firmware.py:86]: I think the doc strings of put with and without UID may be switched? (You update a firmware with a UID and upload one without it)

File: src/web_interface/rest/rest_missing_analyses.py [hunk 7]
```diff
@@ -9,14 +9,24 @@
 from web_interface.security.privileges import PRIVILEGES
 
 
+api = Namespace('rest/missing', description='Search the database for missing entries')
+
+
+@api.route('', doc={'description': ''})
```
Comment (jstucke) [inline src/web_interface/rest/rest_missing_analyses.py:15]: you can remove the parameter `doc` if it is empty

File: src/web_interface/rest/rest_missing_analyses.py [hunk 8]
```diff
@@ -9,14 +9,24 @@
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
         self.config = kwargs.get('config', None)
 
     @roles_accepted(*PRIVILEGES['delete'])
+    @api.doc(responses={200: 'Success', 400: 'Unknown'})
     def get(self):
+        '''
+        Search for missing entries
+        Search for missing files or missing analyses
```
Comment (jstucke) [inline src/web_interface/rest/rest_missing_analyses.py:28]: This endpoint has some additional functionality that is still missing from the REST docs
```suggestion
        Search for missing or orphaned files and missing or failed analyses
```

File: src/web_interface/rest/rest_status.py [hunk 9]
```diff
@@ -7,15 +7,24 @@
 from web_interface.security.decorator import roles_accepted
 from web_interface.security.privileges import PRIVILEGES
 
+api = Namespace('rest/status', description='Request FACT\'s system status')
 
+
+@api.route('', doc={'description': ''})
```
Comment (jstucke) [inline src/web_interface/rest/rest_status.py:13]: see above

File: src/web_interface/rest/rest_status.py [hunk 10]
```diff
@@ -7,15 +7,24 @@
...
 class RestStatus(Resource):
     URL = '/rest/status'
 
-    def __init__(self, **kwargs):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
         self.config = kwargs.get('config', None)
 
     @roles_accepted(*PRIVILEGES['status'])
+    @api.doc(responses={200: 'Success', 400: 'Unknown system status'})
```
Comment (jstucke) [inline src/web_interface/rest/rest_status.py:22]: 'Unknown system status' doesn't really make sense. 'Error' should be fine

**Review by jstucke — CHANGES_REQUESTED** [review]

**Review by dorpvom — APPROVED** [review]

**Review by jstucke — APPROVED** [review]

**Pull request merged**